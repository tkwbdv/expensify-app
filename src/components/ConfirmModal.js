import React from "react"
import Modal from "react-modal";

if (process.env.NODE_ENV !== 'test') Modal.setAppElement('#root'); // prevents error when testing

const ConfirmModal = ({ showModal, onCloseModal, onRemove }) => {
  return (
    <Modal
      isOpen={showModal}
      onRequestClose={onCloseModal}
      overlayClassName="overlay"
      className="modal"
      closeTimeoutMS={200}
      contentLabel="confirm Remove"
    >
      <h3 className="modal__title">Remove this expense?</h3>
      <div className="modal__button-container">
        <button className="button button--secondary button--confirm" onClick={onRemove}>Yes</button>
        <button className="button button--confirm" onClick={onCloseModal}>No</button>
      </div>
    </Modal>
  );
};

export { ConfirmModal as default };