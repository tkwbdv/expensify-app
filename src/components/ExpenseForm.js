import React, { useState } from "react";

import moment from "moment";

import "react-dates/initialize";
import { SingleDatePicker } from 'react-dates';

const ExpenseForm = ({ formData, expense = {} }) => {
  const [description, setDescription] = useState(expense.description || "");
  const [amount, setAmount] = useState((expense.amount / 100) || "");
  const [note, setNote] = useState(expense.note || "");
  const [createdAt, setCreatedAt] = useState(moment(expense.createdAt) || moment());
  const [calendarFocused, setCalendarFocused] = useState(false);
  const [error, setError] = useState("");

  const onAmountChange = (e) => {
    let amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      setAmount(amount);
    }
  };

  const onDescriptionChange = (e) => {
    const description = e.target.value;
    setDescription(description);
  };

  const onNoteChange = (e) => {
    const note = e.target.value;
    setNote(note);
  };

  const onDateChange = (createdAt) => {
    if (createdAt) {
      setCreatedAt(createdAt)
    }
  };

  const onFormSubmit = (e) => {
    e.preventDefault();
    if (!description || !amount) {
      setError("Please provide description and amount.");
    } else {
      setError("");
      formData({
        description,
        amount: +(parseFloat(amount) * 100).toFixed(0),
        createdAt: +createdAt,  // momentObject -> unix timestamp
        note
      });
    }
  };


  return (
    <div>
      {error && <p>{error}</p>}
      <form onSubmit={onFormSubmit}>
        <input
          placeholder="Description"
          autoFocus
          value={description}
          onChange={onDescriptionChange}
        />
        <input
          placeholder="Amount"
          type="text"
          value={amount}
          onChange={onAmountChange}
        />
        <SingleDatePicker
          date={createdAt}
          onDateChange={onDateChange}
          focused={calendarFocused}
          onFocusChange={({ focused }) => setCalendarFocused(focused)}
          id="your_unique_id"
          noBorder={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
          hideKeyboardShortcutsPanel={true}
        />
        <textarea
          placeholder="Add a note for your expense (optional)"
          value={note}
          onChange={onNoteChange}
        />
        <input
          type="submit"
          value={expense.description ? "Update Expense" : "Create Expense"}
        />
      </form>
    </div >
  )
};

export default ExpenseForm;