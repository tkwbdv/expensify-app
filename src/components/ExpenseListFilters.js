import React, { useState } from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from "../actions/filters";
import { DateRangePicker } from 'react-dates';

export const ExpenseListFilters = ({ filters, setStartDate, setEndDate, setTextFilter, sortByAmount, sortByDate }) => {
  const [focusedInput, setFocusedInput] = useState(null);

  const onDatesChange = ({ startDate, endDate }) => {
    setStartDate(startDate);
    setEndDate(endDate);
  };

  const onTextChange = (e) => {
    setTextFilter(e.target.value);
  }

  const onSortChange = (e) => {
    if (e.target.value === "amount") {
      sortByAmount();
    } else if (e.target.value === "date") {
      sortByDate();
    }
  }

  return (
    <div>
      <input
        type="text"
        value={filters.text}
        onChange={onTextChange}
      />
      <select
        value={filters.sortBy}
        onChange={onSortChange}
      >
        <option value="amount">Amount</option>
        <option value="date">Date</option>
      </select>
      <DateRangePicker
        startDate={filters.startDate}
        startDateId="your_unique_start_date_id"
        endDate={filters.endDate}
        endDateId="your_unique_end_date_id"
        onDatesChange={onDatesChange}
        focusedInput={focusedInput}
        onFocusChange={focusedInput => setFocusedInput(focusedInput)}
        hideKeyboardShortcutsPanel={true}
        noBorder={true}
        isOutsideRange={() => false}
        numberOfMonths={1}
        showClearDates={true}
      />
    </div>
  );
};


const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);