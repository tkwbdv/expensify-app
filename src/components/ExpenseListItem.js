import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>{amount} - {moment(createdAt).format("MMMM Do YYYY")}</p>
  </div>
);

export default ExpenseListItem;