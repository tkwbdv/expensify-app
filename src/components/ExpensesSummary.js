import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import getVisibleExpenses from "../selectors/expenses";
import expensesTotal from "../selectors/expenses-total";
import numeral from "numeral";

export const ExpensesSummary = ({ expenseCount, expensesTotal, hiddenExpensesCount }) => {
  const expenseWord = expenseCount === 1 ? "expense" : "expenses";
  const hiddenExpenseWord = hiddenExpensesCount === 1 ? "expense" : "expenses";
  const formattedExpensesTotal = numeral(expensesTotal / 100).format("$0,0.00");
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{expenseCount}</span> {expenseWord} totalling <span>{formattedExpensesTotal}</span>

          {!!hiddenExpensesCount && <div><h6 className="page-header__subtitle"><span>{hiddenExpensesCount}</span> {hiddenExpenseWord} hidden by filters</h6></div>}

          <div className="page-header__actions">
            <Link className="button" to="/create">Add Expense</Link>
          </div>
        </h1>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);
  const hiddenExpensesCount = state.expenses.length - visibleExpenses.length;

  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: expensesTotal(visibleExpenses),
    hiddenExpensesCount
  };
};

export default connect(mapStateToProps)(ExpensesSummary);