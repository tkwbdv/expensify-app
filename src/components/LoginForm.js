import React, { useState } from "react";
import { connect } from "react-redux";
import { startLogin, startSignup } from "../actions/auth";

const LoginForm = ({ startLoginDispatch, startSignupDispatch }) => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState(undefined);

  const onEmailChange = (e) => setEmail(e.target.value);
  const onPasswordChange = (e) => setPassword(e.target.value);

  const handleClickSignup = (e) => {
    e.preventDefault();
    startSignup(email, password)()
      .then(res => typeof res === "string" ? setError(res) : startSignupDispatch(email, password));
  };

  const handleClickLogin = (e) => {
    e.preventDefault();
    startLogin(undefined, email, password)()
      .then(res => typeof res === "string" ? setError(res) : startLoginDispatch(undefined, email, password));
  };

  return (
    <div>
      <h3 className="modal__title">Login via E-Mail</h3>
      <div className="modal__button-container">
      </div>
      <form className="form">
        E-Mail: <input type="text" autoFocus value={email} onChange={onEmailChange} />
        Password: <input type="password" value={password} onChange={onPasswordChange} />
        {error && <p>{error}</p>}
        <div className="modal__button-container">
          <button type="submit" className="button button--confirm" onClick={handleClickLogin}>Login</button>
          <button className="button button--confirm button--secondary" onClick={handleClickSignup}>Signup</button>
        </div>
      </form>
    </div>
  );
};

const MapDispatchToProps = (dispatch) => ({
  startLoginDispatch: (provider, email, password) => dispatch(startLogin(provider, email, password)),
  startSignupDispatch: (email, password) => dispatch(startSignup(email, password))
});

export default connect(undefined, MapDispatchToProps)(LoginForm)