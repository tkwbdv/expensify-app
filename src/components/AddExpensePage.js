import React from "react";
import { connect } from "react-redux";
import { startAddExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";

export const AddExpensePage = ({ history, startAddExpense }) => (
  <div>
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">Add Expense</h1>
      </div>
    </div>
    <div className="content-container">
      <ExpenseForm
        formData={(expense) => {
          startAddExpense(expense);
          history.push("/");
        }}
      />
    </div>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense, dispatch))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage);