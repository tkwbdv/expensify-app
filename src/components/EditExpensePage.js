import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";
import NotFoundPage from "./NotFoundPage";

export const EditExpensePage = ({ match, startEditExpense, startRemoveExpense, expense, history }) => {
  if (!expense) return (
    <NotFoundPage />
  );

  const formData = (expenseData) => {
    startEditExpense(match.params.id, expenseData);
    history.push("/");
  }

  const onRemove = () => {
    startRemoveExpense({ id: match.params.id });
    history.push("/");
  };

  return (
    <div>
      <h1>Edit Expense</h1>
      <ExpenseForm
        formData={formData}
        expense={expense}
      />
      <button onClick={onRemove}>Remove</button>
    </div>
  );
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
  startEditExpense: (id, expenseData) => dispatch(startEditExpense(id, expenseData)),
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);