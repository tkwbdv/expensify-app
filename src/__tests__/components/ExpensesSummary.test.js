import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/ExpensesSummary";

test("should render ExpenseSummary correctly with 0 expenses", () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={0} expensesTotal={0} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseSummary correctly with 1 expense", () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={1500} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseSummary correctly with multiple expenses", () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={2} expensesTotal={2500} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseSummary correctly with 1 filtered expense", () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={2} expensesTotal={2500} hiddenExpenses={1} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseSummary correctly with multiple filtered expense", () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={2} expensesTotal={2500} hiddenExpenses={2} />);
  expect(wrapper).toMatchSnapshot();
});