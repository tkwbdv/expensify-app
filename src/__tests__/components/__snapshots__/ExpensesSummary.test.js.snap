// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary correctly with 0 expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        0
      </span>
       
      expenses
       totalling 
      <span>
        $0.00
      </span>
      <div
        className="page-header__actions"
      >
        <Link
          className="button"
          to="/create"
        >
          Add Expense
        </Link>
      </div>
    </h1>
  </div>
</div>
`;

exports[`should render ExpenseSummary correctly with 1 expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $15.00
      </span>
      <div
        className="page-header__actions"
      >
        <Link
          className="button"
          to="/create"
        >
          Add Expense
        </Link>
      </div>
    </h1>
  </div>
</div>
`;

exports[`should render ExpenseSummary correctly with 1 filtered expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        2
      </span>
       
      expenses
       totalling 
      <span>
        $25.00
      </span>
      <div
        className="page-header__actions"
      >
        <Link
          className="button"
          to="/create"
        >
          Add Expense
        </Link>
      </div>
    </h1>
  </div>
</div>
`;

exports[`should render ExpenseSummary correctly with multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        2
      </span>
       
      expenses
       totalling 
      <span>
        $25.00
      </span>
      <div
        className="page-header__actions"
      >
        <Link
          className="button"
          to="/create"
        >
          Add Expense
        </Link>
      </div>
    </h1>
  </div>
</div>
`;

exports[`should render ExpenseSummary correctly with multiple filtered expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        2
      </span>
       
      expenses
       totalling 
      <span>
        $25.00
      </span>
      <div
        className="page-header__actions"
      >
        <Link
          className="button"
          to="/create"
        >
          Add Expense
        </Link>
      </div>
    </h1>
  </div>
</div>
`;
